src/Compilers/Shared/NamedPipeUtil.cs

```diff
@@ -0,0 +1,162 @@
+        /// </summary>
+        private static PipeOptions CurrentUserOption = PlatformInformation.IsRunningOnMono
+            ? (PipeOptions)s_currentUserOnlyValue
+            : PipeOptions.None;
+
+        internal static NamedPipeServerStream CreateServer(
+            string pipeName,
+            PipeDirection direction,
+            int maxNumberOfServerInstances,
+            PipeTransmissionMode transmissionMode,
+            PipeOptions options,
+            int inBufferSize,
+            int outBufferSize) =>
+            new NamedPipeServerStream(
+                pipeName,
+                direction,
+                maxNumberOfServerInstances,
+                transmissionMode,
+                options | CurrentUserOption,
+                inBufferSize,
+                outBufferSize,
+                CreatePipeSecurity(),
+                HandleInheritability.None);
+
+        /// <summary>
+        /// Check to ensure that the named pipe server we connected to is owned by the same
+        /// user.
+        /// </summary>
+        internal static bool CheckPipeConnectionOwnership(NamedPipeClientStream pipeStream)
+        {
+            if (PlatformInformation.IsWindows)
+            {
+                var currentIdentity = WindowsIdentity.GetCurrent();
+                var currentOwner = currentIdentity.Owner;
+                var remotePipeSecurity = pipeStream.GetAccessControl();
+                var remoteOwner = remotePipeSecurity.GetOwner(typeof(SecurityIdentifier));
+                return currentOwner.Equals(remoteOwner);
+            }
+
+            // Client side validation isn't supported on Unix. The model relies on the server side
+            // security here.
+            return true;
+        }
+
+        internal static PipeSecurity CreatePipeSecurity()
+        {
+            if (PlatformInformation.IsRunningOnMono)
+            {
+                // Pipe security and additional access rights constructor arguments
+                //  are not supported by Mono 
+                // https://github.com/dotnet/roslyn/pull/30810
+                // https://github.com/mono/mono/issues/11406
+                return null;
+            }
+
+            var security = new PipeSecurity();
+            SecurityIdentifier identifier = WindowsIdentity.GetCurrent().Owner;
+
+            // Restrict access to just this account.  
+            PipeAccessRule rule = new PipeAccessRule(identifier, PipeAccessRights.ReadWrite | PipeAccessRights.CreateNewInstance, AccessControlType.Allow);
+            security.AddAccessRule(rule);
+            security.SetOwner(identifier);
+            return security;
+        }
+
+#elif NETCOREAPP2_1
+
+        private static PipeOptions CurrentUserOption = PipeOptions.CurrentUserOnly;
+
+        // Validation is handled by CurrentUserOnly
+        internal static bool CheckPipeConnectionOwnership(NamedPipeClientStream pipeStream) => true;
```

Comment:
In general it seems cleaner to if-def each function individually so you can share the signature for all the methods.

Reply:
I initially did it that way but I felt like the file was completely un-readable. There are three sections per function and it made the code quite messy. This way felt a bit more readable.